rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,410 @@
+2. Query the `/v1/webapi/find` endpoint.
+3. Check that `agent_auto_updates` is true.
+4. If the current version of Teleport is the latest, quit.
+5. Wait `random(0, agent_update_jitter_seconds)` seconds.
+6. Ensure there is enough free disk space to upgrade Teleport.
+7. Download the desired Teleport tarball specified by `agent_version` and `server_edition`.
+8. Download and verify the checksum (tarball URL suffixed with `.sha256`).
+9. Extract the tarball to `/var/lib/teleport/versions/VERSION`.
+10. Update symlinks to point at the new version.
+11. Backup `/var/lib/teleport/proc/sqlite.db` into `/var/lib/teleport/versions/OLD-VERSION/backup/sqlite.db` and create `backup.yaml`.
+12. Restart the agent if the systemd service is already enabled.
+13. Set `active_version` in `updates.yaml` if successful or not enabled.
+14. Replace the old symlinks/binaries and `/var/lib/teleport/proc/sqlite.db` and quit (exit 1) if unsuccessful.
+15. Remove all stored versions of the agent except the current version and last working version.
+
+To enable auto-updates of the updater itself, all commands will first check for an `active_version`, and reexec using the `teleport-updater` at that version if present and different.
+The `/usr/local/bin/teleport-updater` symlink will take precedence to avoid reexec in most scenarios.
+
+To ensure that SELinux permissions do not prevent the `teleport-updater` binary from installing/removing Teleport versions, the updater package will configure SELinux contexts to allow changes to all required paths.
+
+To ensure that `teleport` package removal does not interfere with `teleport-updater`, package removal will run `apt purge` (or `yum` equivalent) while ensuring that `/etc/teleport.yaml` and `/var/lib/teleport` are not purged.
+Failure to do this could result in `/etc/teleport.yaml` being removed when an operator runs `apt purge` at a later date.
+
+To ensure that backups are consistent, the updater will use the [SQLite backup API](https://www.sqlite.org/backup.html) to perform the backup.
+
+#### Failure Conditions
+
+If the new version of Teleport fails to start, the installation of Teleport is reverted as described above.
+
+If `teleport-updater` itself fails with an error, and an older version of `teleport-updater` is available, the upgrade will retry with the older version.
+
+Known failure conditions caused by intentional configuration (e.g., upgrades disabled) will not trigger retry logic.
+
+#### Status
+
+To retrieve known information about agent upgrades, the `status` subcommand will return the following:
+```json
+{
+  "agent_version_installed": "15.1.1",
+  "agent_version_desired": "15.1.2",
+  "agent_version_previous": "15.1.0",
+  "agent_edition_installed": "enterprise",
+  "agent_edition_desired": "enterprise",
+  "agent_edition_previous": "enterprise",
+  "agent_update_time_last": "2020-12-10T16:00:00+00:00",
+  "agent_update_time_jitter": 600,
+  "agent_updates_enabled": true
+}
+```
+
+### Downgrades
```

Comment:
@sclevine @espadolini Should we backup `teleport.yaml` as well? That would protect us against schema changes in the configuration file as well.

Reply:
Do Teleport agents re-write `teleport.yaml` when upgraded? If so, I agree

Reply:
Good point, I don't think we need to back it up.